File: matrix/diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
         if (mat.empty() || mat[0].empty())
            return {};

        int rows = mat.size();
        int cols = mat[0].size();
        vector<int> res;
        res.reserve(rows * cols);  // Pre-allocate for performance

        int r = 0, c = 0;
        bool up = true;

        for (int i = 0; i < rows * cols; ++i) {
            res.push_back(mat[r][c]);

            if (up) {
                if (c == cols - 1) {
                    ++r;
                    up = false;
                } else if (r == 0) {
                    ++c;
                    up = false;
                } else {
                    --r;
                    ++c;
                }
            } else {
                if (r == rows - 1) {
                    ++c;
                    up = true;
                } else if (c == 0) {
                    ++r;
                    up = true;
                } else {
                    ++r;
                    --c;
                }
            }
        }

        return res;
    }
};

File: simulation/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if not mat or not mat[0]: return []

        rows, cols = len(mat), len(mat[0])
        result = []
        r, c = 0, 0
        up = True

        for _ in range(rows * cols):
            result.append(mat[r][c])

            if up:
                if c == cols - 1:
                    r += 1
                    up = False
                elif r == 0:
                    c += 1
                    up = False
                else:
                    r -= 1
                    c += 1
            else:
                if r == rows - 1:
                    c += 1
                    up = True
                elif c == 0:
                    r += 1
                    up = True
                else:
                    r += 1
                    c -= 1

        return result